app/src/main/java/com/example/mystore/ui/home/HomeViewModel.kt

```diff
@@ -16,33 +20,106 @@ import javax.inject.Inject
 class HomeViewModel @Inject constructor(
     private val apiRepository: IApiRepository
 ) : ViewModel() {
+    private var allProducts = mutableStateListOf<ProductModel>()
     var products = mutableStateListOf<ProductModel>()
     var isLoading by mutableStateOf(false)
+    var categories = mutableStateListOf<CategoryDetailsModel>()
+
+    var input by mutableStateOf("")
+    private var priceStartRange by mutableFloatStateOf(20f)
+    private var priceEndRange by mutableFloatStateOf(150f)
+    private var rating by mutableIntStateOf(4)
+    private var selectedCategories = mutableStateListOf<String>().apply {
+        addAll(listOf("Home"))
+    }
+    var filtersCount by mutableIntStateOf(3)
 
     init {
         getProducts()
+        getCategories()
+    }
+
+    fun filterProducts(newInput: String? = null) {
+        products.apply {
+            clear()
+            addAll(allProducts.filter {
+                it.price in (priceStartRange..priceEndRange) && it.rating <= rating
+            })
+        }
+
+        if (selectedCategories.isNotEmpty() && selectedCategories.count() != categories.count()) {
+            products.apply {
+                removeAll(products.filter {
+                    !selectedCategories.contains(it.category)
+                })
+            }
+        }
+
+        if (newInput != null) {
+            input = newInput
+        }
+        if (input.isNotEmpty()) {
+            products.apply {
+                removeAll(
+                    products.filter { product ->
+                        !(product.title.lowercase().contains(input.lowercase()) ||
```

Comment:
This is a bit long. I'd personally write a String extension function that checks if a lowercased string is contained in the lowercased `this` string and then using it here will make the code a lot more legible.